dm/pkg/checker/privilege.go

```diff
@@ -122,89 +143,214 @@ func (pc *SourceReplicatePrivilegeChecker) Name() string {
 	return "source db replication privilege checker"
 }
 
-// TODO: if we add more privilege in future, we might add special checks (globally granted?) for that new privilege.
-func verifyPrivileges(result *Result, grants []string, expectedGrants map[mysql.PrivilegeType]struct{}) {
-	result.State = StateFailure
+func verifyPrivileges(result *Result, grants []string, lackPriv map[mysql.PrivilegeType]map[string]map[string]struct{}) *Error {
 	if len(grants) == 0 {
-		result.Errors = append(result.Errors, NewError("there is no such grant defined for current user on host '%%'"))
-		return
-	}
-
-	var (
-		user       string
-		lackGrants = make(map[mysql.PrivilegeType]struct{}, len(expectedGrants))
-	)
-	for k := range expectedGrants {
-		lackGrants[k] = struct{}{}
+		return NewError("there is no such grant defined for current user on host '%%'")
 	}
 
+	p := parser.New()
 	for i, grant := range grants {
-		// get username and hostname
-		node, err := parser.New().ParseOneStmt(grant, "", "")
+		if len(lackPriv) == 0 {
+			break
+		}
+		node, err := p.ParseOneStmt(grant, "", "")
 		if err != nil {
-			result.Errors = append(result.Errors, NewError(errors.Annotatef(err, "grant %s, grant after replace %s", grants[i], grant).Error()))
-			return
+			return NewError(errors.Annotatef(err, "grant %s, grant after replace %s", grants[i], grant).Error())
 		}
 		grantStmt, ok := node.(*ast.GrantStmt)
 		if !ok {
 			switch node.(type) {
 			case *ast.GrantProxyStmt, *ast.GrantRoleStmt:
 				continue
 			default:
-				result.Errors = append(result.Errors, NewError("%s is not grant statement", grants[i]))
-				return
+				return NewError("%s is not grant statement", grants[i])
 			}
 		}
 
 		if len(grantStmt.Users) == 0 {
-			result.Errors = append(result.Errors, NewError("grant has no user %s", grantStmt.Text()))
-			return
-		} else if user == "" {
-			// show grants will only output grants for requested user
-			user = grantStmt.Users[0].User.Username
+			return NewError("grant has no user %s", grantStmt.Text())
 		}
 
-		for _, privElem := range grantStmt.Privs {
-			if privElem.Priv == mysql.AllPriv {
-				if grantStmt.Level.Level == ast.GrantLevelGlobal {
-					result.State = StateSuccess
-					return
-				}
-				// REPLICATION CLIENT, REPLICATION SLAVE, RELOAD should be global privileges,
-				// thus a non-global GRANT ALL is not enough
-				for expectedGrant := range lackGrants {
-					if _, ok := privNeedGlobal[expectedGrant]; !ok {
-						delete(lackGrants, expectedGrant)
+		dbName := grantStmt.Level.DBName
+		tableName := grantStmt.Level.TableName
+		switch grantStmt.Level.Level {
+		case ast.GrantLevelGlobal:
+			for _, privElem := range grantStmt.Privs {
+				if privElem.Priv == mysql.AllPriv {
+					if _, ok := lackPriv[mysql.GrantPriv]; ok {
+						lackPriv = make(map[mysql.PrivilegeType]map[string]map[string]struct{})
```

Comment:
please explain why we reset lackPriv?

Reply:
L196-197 have its reason, I will copy it here.